File: frontend/src/app/services/team.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, BehaviorSubject } from 'rxjs';
import { tap } from 'rxjs/operators';

export interface Team {
  id: string;
  name: string;
  description: string;
  leader_id: string;
  member_ids: string[];
  invite_code: string;
  score: number;
  created_at: string;
  updated_at: string;
}

export interface TeamMember {
  id: string;
  username: string;
  is_leader: boolean;
}

export interface TeamInvitation {
  id: string;
  team_id: string;
  team_name: string;
  inviter_id: string;
  inviter_name: string;
  invitee_email?: string;
  invitee_user_id?: string;
  status: string;
  expires_at: string;
  created_at: string;
}

export interface TeamResponse {
  message: string;
  team?: Team;
  members?: TeamMember[];
}

export interface InvitationsResponse {
  invitations: TeamInvitation[];
}

export interface TeamScoreboardResponse {
  teams: Team[];
}

@Injectable({
  providedIn: 'root'
})
export class TeamService {
  private apiUrl = 'http://localhost:8080/teams';
  private currentTeamSubject = new BehaviorSubject<Team | null>(null);
  
  currentTeam$ = this.currentTeamSubject.asObservable();

  constructor(private http: HttpClient) {
    // Check team status on init
    this.checkTeamStatus();
  }

  // Team CRUD operations
  createTeam(name: string, description: string): Observable<TeamResponse> {
    return this.http.post<TeamResponse>(`${this.apiUrl}`, { name, description }, { withCredentials: true }).pipe(
      tap(response => {
        if (response.team) {
          this.currentTeamSubject.next(response.team);
        }
      })
    );
  }

  getMyTeam(): Observable<TeamResponse> {
    return this.http.get<TeamResponse>(`${this.apiUrl}/my-team`, { withCredentials: true }).pipe(
      tap(response => {
        if (response.team) {
          this.currentTeamSubject.next(response.team);
        }
      })
    );
  }

  getTeamDetails(teamId: string): Observable<TeamResponse> {
    return this.http.get<TeamResponse>(`${this.apiUrl}/${teamId}`, { withCredentials: true });
  }

  updateTeam(teamId: string, name: string, description: string): Observable<TeamResponse> {
    return this.http.put<TeamResponse>(`${this.apiUrl}/${teamId}`, { name, description }, { withCredentials: true }).pipe(
      tap(response => {
        if (response.team) {
          this.currentTeamSubject.next(response.team);
        }
      })
    );
  }

  deleteTeam(teamId: string): Observable<TeamResponse> {
    return this.http.delete<TeamResponse>(`${this.apiUrl}/${teamId}`, { withCredentials: true }).pipe(
      tap(() => {
        this.currentTeamSubject.next(null);
      })
    );
  }

  // Invitation methods
  inviteByUsername(teamId: string, username: string): Observable<any> {
    return this.http.post<any>(`${this.apiUrl}/${teamId}/invite/username`, { username }, { withCredentials: true });
  }

  inviteByEmail(teamId: string, email: string): Observable<any> {
    return this.http.post<any>(`${this.apiUrl}/${teamId}/invite/email`, { email }, { withCredentials: true });
  }

  joinByCode(code: string): Observable<TeamResponse> {
    return this.http.post<TeamResponse>(`${this.apiUrl}/join/${code}`, {}, { withCredentials: true }).pipe(
      tap(response => {
        if (response.team) {
          this.currentTeamSubject.next(response.team);
        }
      })
    );
  }

  getPendingInvitations(): Observable<InvitationsResponse> {
    return this.http.get<InvitationsResponse>(`${this.apiUrl}/invitations`, { withCredentials: true });
  }

  acceptInvitation(invitationId: string): Observable<TeamResponse> {
    return this.http.post<TeamResponse>(`${this.apiUrl}/invitations/${invitationId}/accept`, {}, { withCredentials: true }).pipe(
      tap(response => {
        if (response.team) {
          this.currentTeamSubject.next(response.team);
        }
      })
    );
  }

  rejectInvitation(invitationId: string): Observable<any> {
    return this.http.post<any>(`${this.apiUrl}/invitations/${invitationId}/reject`, {}, { withCredentials: true });
  }

  // Team invitation management (for leaders)
  getTeamPendingInvitations(teamId: string): Observable<InvitationsResponse> {
    return this.http.get<InvitationsResponse>(`${this.apiUrl}/${teamId}/invitations`, { withCredentials: true });
  }

  cancelInvitation(teamId: string, invitationId: string): Observable<any> {
    return this.http.delete<any>(`${this.apiUrl}/${teamId}/invitations/${invitationId}`, { withCredentials: true });
  }

  // Member management
  removeMember(teamId: string, userId: string): Observable<any> {
    return this.http.delete<any>(`${this.apiUrl}/${teamId}/members/${userId}`, { withCredentials: true });
  }

  leaveTeam(teamId: string): Observable<any> {
    return this.http.post<any>(`${this.apiUrl}/${teamId}/leave`, {}, { withCredentials: true }).pipe(
      tap(() => {
        this.currentTeamSubject.next(null);
      })
    );
  }

  regenerateInviteCode(teamId: string): Observable<any> {
    return this.http.post<any>(`${this.apiUrl}/${teamId}/regenerate-code`, {}, { withCredentials: true });
  }

  // Scoreboard
  getTeamScoreboard(): Observable<TeamScoreboardResponse> {
    return this.http.get<TeamScoreboardResponse>('http://localhost:8080/scoreboard/teams', { withCredentials: true });
  }

  // Helper methods
  getCurrentTeam(): Team | null {
    return this.currentTeamSubject.value;
  }

  isInTeam(): boolean {
    return this.currentTeamSubject.value !== null;
  }

  isTeamLeader(): boolean {
    // This requires user ID from auth service
    return false; // Will be implemented with proper check
  }

  checkTeamStatus(): void {
    this.http.get<TeamResponse>(`${this.apiUrl}/my-team`, { withCredentials: true }).subscribe({
      next: (response) => {
        if (response.team) {
          this.currentTeamSubject.next(response.team);
        } else {
          this.currentTeamSubject.next(null);
        }
      },
      error: () => {
        this.currentTeamSubject.next(null);
      }
    });
  }
}

File: backend/internal/models/team.go
package models

import (
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

// Team represents a team in the CTF competition
type Team struct {
	ID          primitive.ObjectID   `bson:"_id,omitempty" json:"id"`
	Name        string               `bson:"name" json:"name"`
	Description string               `bson:"description" json:"description"`
	LeaderID    primitive.ObjectID   `bson:"leader_id" json:"leader_id"`
	MemberIDs   []primitive.ObjectID `bson:"member_ids" json:"member_ids"`
	InviteCode  string               `bson:"invite_code" json:"invite_code"`
	Score       int                  `bson:"score" json:"score"`
	CreatedAt   time.Time            `bson:"created_at" json:"created_at"`
	UpdatedAt   time.Time            `bson:"updated_at" json:"updated_at"`
}

// TeamInvitation represents an invitation to join a team
type TeamInvitation struct {
	ID            primitive.ObjectID `bson:"_id,omitempty" json:"id"`
	TeamID        primitive.ObjectID `bson:"team_id" json:"team_id"`
	TeamName      string             `bson:"team_name" json:"team_name"`
	InviterID     primitive.ObjectID `bson:"inviter_id" json:"inviter_id"`
	InviterName   string             `bson:"inviter_name" json:"inviter_name"`
	InviteeEmail  string             `bson:"invitee_email,omitempty" json:"invitee_email,omitempty"`
	InviteeUserID primitive.ObjectID `bson:"invitee_user_id,omitempty" json:"invitee_user_id,omitempty"`
	Token         string             `bson:"token" json:"token"`
	Status        string             `bson:"status" json:"status"` // pending, accepted, rejected, expired
	ExpiresAt     time.Time          `bson:"expires_at" json:"expires_at"`
	CreatedAt     time.Time          `bson:"created_at" json:"created_at"`
}

// Invitation statuses
const (
	InvitationStatusPending  = "pending"
	InvitationStatusAccepted = "accepted"
	InvitationStatusRejected = "rejected"
	InvitationStatusExpired  = "expired"
)

// Team size constants
const (
	MinTeamSize = 2
	MaxTeamSize = 4
)
